redis_commands.c

```diff
@@ -5743,52 +5743,89 @@ int redis_xclaim_cmd(INTERNAL_FUNCTION_PARAMETERS, RedisSock *redis_sock,
 }
 
 /* XGROUP HELP
- * XGROUP CREATE key groupname id [MKSTREAM]
- * XGROUP SETID key group id
- * XGROUP DESTROY key groupname
- * XGROUP DELCONSUMER key groupname consumername */
+ * XGROUP CREATE          key group id       [MKSTREAM] [ENTRIESREAD <n>]
+ * XGROUP SETID           key group id                  [ENTRIESREAD <n>]
+ * XGROUP CREATECONSUMER  key group consumer
+ * XGROUP DELCONSUMER     key group consumer
+ * XGROUP DESTROY         key group
+ */
 int redis_xgroup_cmd(INTERNAL_FUNCTION_PARAMETERS, RedisSock *redis_sock,
                      char **cmd, int *cmd_len, short *slot, void **ctx)
 {
-    char *op, *key = NULL, *arg1 = NULL, *arg2 = NULL;
-    size_t oplen, keylen, arg1len, arg2len;
-    zend_bool mkstream = 0;
+    #define REDIS_HAVE_MKSTREAM    0x01
+    #define REDIS_HAVE_ENTRIESREAD 0x02
+
+    zend_string *op = NULL, *key = NULL, *group = NULL, *id_or_consumer = NULL;
+    zend_long entries_read = -2;
     int argc = ZEND_NUM_ARGS();
+    smart_string cmdstr = {0};
+    zend_bool mkstream = 0;
 
-    if (zend_parse_parameters(ZEND_NUM_ARGS(), "s|sssb", &op, &oplen,
-                              &key, &keylen, &arg1, &arg1len, &arg2, &arg2len,
-                              &mkstream) == FAILURE)
-    {
+    /* Ephemeral struct about the args we need for each supported subcommand */
+    struct xgroup_op { const char *str; size_t len; int nargs; int flags; } xgroup_ops[] = {
+        {ZEND_STRL("HELP"), 0, 0},
+        {ZEND_STRL("CREATE"), 3, REDIS_HAVE_MKSTREAM | REDIS_HAVE_ENTRIESREAD },
+        {ZEND_STRL("SETID"), 3, REDIS_HAVE_ENTRIESREAD },
+        {ZEND_STRL("CREATECONSUMER"), 3, 0},
+        {ZEND_STRL("DELCONSUMER"), 3, 0},
+        {ZEND_STRL("DESTROY"), 2, 0}
+    }, *xg_op = NULL;
+
+    ZEND_PARSE_PARAMETERS_START(1, 6)
+        Z_PARAM_STR(op)
+        Z_PARAM_OPTIONAL
+        Z_PARAM_STR(key)
+        Z_PARAM_STR(group)
+        Z_PARAM_STR(id_or_consumer)
+        Z_PARAM_BOOL(mkstream)
+        Z_PARAM_LONG(entries_read)
+    ZEND_PARSE_PARAMETERS_END_EX(return FAILURE);
+
+    /* Find the subcommand */
+    for (size_t i = 0; i < sizeof(xgroup_ops)/sizeof(xgroup_ops[0]); i++) {
```

Comment:
IMO six `if/else` statements would be more obvious and simple and don't need ephemeral structs...

Reply:
OK I removed my cool inline structure 😭 

(I like this better too)